Update Algebra Homomorphisms

raulepure opened the pull request:
Changed AlgHom type by replacing the lex ordering in the computations by a dp*dp resp. dp*lex ordering.
Added groebner_assure with product orderings.
Replaced mapping of elements via Singular.jl by using MapFromFunction and evaluate.
Added check of homoegeneity in case codomain is homogeneous.

Commit:
Update Algebra Homomorphisms

Changed AlgHom type by replacing the lex ordering in the computations by a dp*dp resp. dp*lex ordering.
Added groebner_assure with product orderings.
Replaced mapping of elements via Singular.jl by using MapFromFunction and evaluate.
Added check of homoegeneity in case codomain is homogeneous.

Commit:
Removed commented lines and a useless if check.

tthsqe12 commented:
Working on it. related https://github.com/oscar-system/Oscar.jl/pull/354

tthsqe12 commented on src/Rings/affine-algebra-homs.jl:
```diff
@@ -84,20 +84,69 @@ mutable struct AlgHom{T} <: AbstractAlgebra.Map{Ring, Ring,
-      z = new(R, S, V, Singular.AlgebraHomomorphism(Rx, Sx, Sx.(V)))
+      if typeof(C) <: MPolyRing_dec
+         bit = ishomogenous.(V)
+         length(findall(x->x==1, bit)) == length(V) || error("Array needs to contain homogeneous elements")
```
Isn't this the same thing as `all(ishomogenous, V) || error(...)`

tthsqe12 replied:
```suggestion
         all(ishomogenous, V) || error("Array needs to contain homogeneous elements")
```

tthsqe12 commented on src/Rings/affine-algebra-homs.jl:
```diff
@@ -84,20 +84,69 @@ mutable struct AlgHom{T} <: AbstractAlgebra.Map{Ring, Ring,
+      
+      if typeof(C) <: MPolyRing_dec
+         bit = ishomogenous.(V)
+         length(findall(x->x==1, bit)) == length(V) || error("Array needs to contain homogeneous elements") 
```
ditto

tthsqe12 replied:
```suggestion
         all(ishomogenous, V) || error("Array needs to contain homogeneous elements")
```

tthsqe12 reviewed (commented):
There is a lot here, and I can just make trivial comments at this time. I will try to go more indepth tomorrow.

tthsqe12 commented:
Do the barriers to makeing AlgHom parametrized by the domain and codomain element types still exist?

raulepure commented:
There are still some parts which are problematic like using the base_ring function on an quotient ring and on a graded ring. In one case you get the underlying polynomial ring in the other case you get the base field. 
And I also had some trouble with evaluate in this context,.
This yields to the separation of these types in the code. 

If it is what you mean.

thofma commented:
(We can now use `coefficient_ring` to get the underling coefficients for graded rings. Not sure if this helps though.)

tthsqe12 commented:
Currently the `T` in `AlgHom{T}` is the type of the coefficients in the domain and codomain. This makes writing code against `AlgHom` more difficult, and, more importantly, omits maps from `Z[x]` to `F_5[z]`. As long as all of the cases that are supposed to work are tested in the test code, I would say this can be merged, and we can extent the `AlgHom` in a later pr using the new `coefficient_ring` standard. Also, unless I am missing something,
```julia
bit = ishomogenous.(V)
length(findall(x->x==1, bit)) == length(V)
```
is a much more confusing way of writing
```julia
all(ishomogenous, V)
```

thofma committed:
Apply suggestions from code review

Co-authored-by: tthsqe12 <tthsqe12@gmail.com>

jieaosong commented:
Just want to comment on the spelling of `ishomogeneous`...

thofma commented:
Ah yes, thanks. Obviously this code is broken since `ishomogenous` does not exist. Can you fix the spelling and add a test @tthsqe12?

raulepure commented:
Thank you for changing the code. @tthsqe12 @thofma 
I fixed another "bit ..." into "all(ishomogeneous, V)" which was overlooked and I added a small test. 
The PR is coming in some minutes, I'm just running the test
